### PR TITLE
Stochastic Constraint Integrate and Fire (SCIF) neuron model for constraint satisfaction problems

### PR DESCRIPTION
**Pull request by srrisbud**
Issue Number: 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [ ] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/lava/proc/scif/process.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        shape: Tuple
+            shape of the sigma process. Default is (1,).
+        bias: int
+            bias current driving the SCIF neuron. Default is 1 (arbitrary).
+        theta: int
+            threshold above which a SCIF neuron would fire winner-take-all
+            spike. Default is 4 (arbitrary).
+        neg_tau_ref: int
+            refractory time period (in number of algorithmic time-steps) for a
+            SCIF neuron after firing winner-take-all spike. Default is -5 (
```
Comment (phstratmann) [inline src/lava/proc/scif/process.py:37]: put "(" into next line

File: src/lava/proc/scif/process.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    def __init__(
+            self,
+            *,
+            shape: ty.Tuple[int, ...],
+            bias: ty.Optional[int] = 1,
+            theta: ty.Optional[int] = 4,
+            neg_tau_ref: ty.Optional[int] = -5) -> None:
+        """
+        Stochastic Constraint Integrate and Fire neuron Process.
+
```
Comment (phstratmann) [inline src/lava/proc/scif/process.py:25]: Maybe worth to mention its update equations here?

File: src/lava/proc/scif/models.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        a_in_data = self.a_in.recv()
+
+        # Define spike vectors
+        s_sig = np.zeros_like(self.v)
+        s_wta = np.zeros_like(self.v)
+
+        # Saturated add the incoming activation
+        self.u += a_in_data
+        self.u[self.u > 2 ** 23 - 1] = 2 ** 23 - 1
+        self.u[self.u < -2 ** 23] = -2 ** 23
```
Comment (phstratmann) [inline src/lava/proc/scif/models.py:47]: np.clip may be faster?

File: src/lava/proc/scif/models.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+        # Populate the buffer for local computation
+        lsb = self.beta.copy()
+        lsb &= 3
+        self.beta <<= 2
+        self.beta[self.beta >= 256] = 0  # Overflow 8-bit unsigned beta to 0
```
Comment (phstratmann) [inline src/lava/proc/scif/models.py:53]: Is this what's happening on Loihi? Or does Loihi perform a modulo operation?

File: src/lava/proc/scif/models.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+        # Gather spike and unsatisfied indices for summation axons
```
Comment (phstratmann) [inline src/lava/proc/scif/models.py:55]: Could you give a bit more detail in this comment? I'm not fully sure why this is being done. 
Alternatively, you may want to add the full equation to the Process docstring.
Thanks

File: src/lava/proc/scif/models.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        lfsr = np.zeros_like(v_to_intg)
+        if lfsr.size > 0:
+            rand_nums = \
+                np.random.randint(-2 ** 15, 2 ** 15 - 1,
+                                  size=np.count_nonzero(self.enable_noise == 1))
+            lfsr[self.enable_noise == 1] = rand_nums
+
+        lfsr = np.right_shift(lfsr, 1)
+        v_to_intg = v_to_intg + lfsr + u_to_intg + bias_to_intg
+        v_to_intg[v_to_intg > 2 ** 23 - 1] = 2 ** 23 - 1  # Saturate at max
```
Comment (phstratmann) [inline src/lava/proc/scif/models.py:96]: np.clip?

File: tests/lava/proc/scif/test_process.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        """Tests instantiation of SigmaDelta"""
+        scif = SCIF(shape=(10,),
+                    bias=2,
+                    theta=8,
+                    neg_tau_ref=-10)
+
+        self.assertEqual(scif.shape, (10,))
+        self.assertEqual(scif.bias.init, 2)
+        self.assertEqual(scif.theta.init, 8)
+        self.assertEqual(scif.neg_tau_ref.init, -10)
```
Comment (phstratmann) [inline tests/lava/proc/scif/test_process.py:32]: empty line missing

File: tests/lava/proc/scif/test_models.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+            np.array(volts_lif_wta).astype(int), \
+            np.array(volts_lif_sig).astype(int)
+
+    def test_scif_fixed_pt_no_noise(self) -> None:
+        """Test a single SCIF neuron without noise, but with a constant bias.
+        The neuron is expected to spike with a regular period, on WTA as well as
+        Sigma axons. After excitatory spikes on two consecutive time-steps, the
+        neuron goes into inhibition and sends 2 inhibitory spikes of payload -1
+        at the end of its refractory period.
+        """
```
Comment (phstratmann) [inline tests/lava/proc/scif/test_models.py:115]: Awesome, extensive docstring :-)

File: tests/lava/proc/scif/test_models.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        num_epochs = 10
+        num_steps = num_epochs * total_period + (theta // bias)
+        v_scif, v_lif_wta, v_lif_sig = self.run_test(num_steps=num_steps,
+                                                     num_neurons=num_neurons,
+                                                     bias=bias,
+                                                     theta=theta,
+                                                     neg_tau_ref=neg_tau_ref,
+                                                     wt=wt,
+                                                     t_inj_spk={})
+        # voltages = np.hstack((np.arange(num_steps).reshape(num_steps, 1),
```
Comment (phstratmann) [inline tests/lava/proc/scif/test_models.py:131]: Maybe delete this comment?

**Review by phstratmann — APPROVED** [review]
Was a pleasure to review your code: easy to read, well documented, well tested.
Only very minor style suggestions. Feel free to implement or ignore.

File: tests/lava/proc/scif/test_models.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+from lava.proc.scif.process import SCIF
+from lava.proc.lif.process import LIF
+from lava.proc.dense.process import Dense
+from lava.proc.io.source import RingBuffer as SpikeSource
+
+verbose = True if (('-v' in sys.argv) or ('--verbose' in sys.argv)) else False
+
+
+class TestSCIFModels(unittest.TestCase):
+    """Tests for sigma delta neuron"""
```
Comment (GaboFGuerra) [inline tests/lava/proc/scif/test_models.py:32]: Docstring seems to belong to tests for other process.

**Review by GaboFGuerra — APPROVED** [review]
Thanks Sumedh! just remove the zombie code PS mentioned in test_models.py.

**Pull request merged**